image-loader/app/lib/Downloader.scala

```diff
@@ -17,42 +17,40 @@ case object TruncatedDownload extends Exception
 case object InvalidDownload extends Exception
 
 //TODO Revisit this logic
-class Downloader(implicit ec: ExecutionContext) extends GridLogging {
-  private val client = new OkHttpClient()
+class Downloader(implicit ec: ExecutionContext, wsClient: WSClient) extends GridLogging {
   private val digester = DeprecatedHashWrapper.sha1()
 
-  def download(uri: URI, file: File): Future[DigestedFile] = Future {
-    val request = new Request.Builder().url(uri.toString).build()
-    val response = client.newCall(request).execute()
+  def download(uri: URI, file: File): Future[DigestedFile] = for {
+    response <- wsClient.url(uri.toString).get()
 
-    val maybeExpectedSize = Try{response.header("Content-Length").toInt}
-
-    maybeExpectedSize match {
-      case Failure(exception) => {
-        logger.error(s"Missing content-length header from $uri", exception)
-        throw InvalidDownload
-      }
-      case Success(expectedSize) => {
-        val input = response.body().byteStream()
-
-        val output = new FileOutputStream(file)
-        val hashedOutput = new HashingOutputStream(digester, output)
+    maybeExpectedSize = Try {
+      response.header("Content-Length").map(_.toInt)
+    }
+  } yield maybeExpectedSize match {
+    case Success(None) =>
+      logger.error(s"Missing content-length header from $uri")
+      throw InvalidDownload
+    case Failure(exception) =>
+      logger.error(s"Bad content-length header from $uri", exception)
+      throw InvalidDownload
+    case Success(Some(expectedSize)) =>
+      val input: java.io.InputStream = new java.io.ByteArrayInputStream(response.bodyAsBytes.toArray)
 
-        ByteStreams.copy(input, hashedOutput)
+      val output = new FileOutputStream(file)
+      val hashedOutput = new HashingOutputStream(digester, output)
 
-        val hash = hashedOutput.hash().asBytes()
+      ByteStreams.copy(input, hashedOutput)
 
-        input.close()
-        hashedOutput.close()
+      val hash = hashedOutput.hash().asBytes()
 
-        val actualSize = Files.size(file.toPath)
+      hashedOutput.close()
 
-        if (actualSize != expectedSize) {
-          throw TruncatedDownload
-        }
+      val actualSize = Files.size(file.toPath)
 
-        DigestedFile(file, hash)
+      if (actualSize != expectedSize) {
+        throw TruncatedDownload
```

Comment:
should this delete the file since it is corrupt?

Reply:
No, it's always deleted in the calling code during onComplete.